core/trino-main/src/main/java/io/trino/sql/planner/optimizations/DetermineHashPartitionCount.java

```diff
@@ -0,0 +1,226 @@
+ * Join output row count: 5000
+ * Estimated hash partition count: max((Input table row count / MIN_INPUT_ROWS_PER_TASK), (Join output row count / MIN_INPUT_ROWS_PER_TASK))  => 10
+ */
+public class DetermineHashPartitionCount
+        implements PlanOptimizer
+{
+    private static final Logger log = Logger.get(DetermineHashPartitionCount.class);
+    private static final List<Class<? extends PlanNode>> SKIP_PLAN_NODES = ImmutableList.of(TableExecuteNode.class, TableWriterNode.class, MergeWriterNode.class);
+
+    private final StatsCalculator statsCalculator;
+
+    public DetermineHashPartitionCount(StatsCalculator statsCalculator)
+    {
+        this.statsCalculator = requireNonNull(statsCalculator, "statsCalculator is null");
+    }
+
+    @Override
+    public PlanNode optimize(
+            PlanNode plan,
+            Session session,
+            TypeProvider types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator,
+            WarningCollector warningCollector,
+            TableStatsProvider tableStatsProvider)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(tableStatsProvider, "tableStatsProvider is null");
+
+        // Skip for write nodes since writing partitioned data with small amount of nodes could cause
+        // memory related issues even when the amount of data is small.
+        if (PlanNodeSearcher.searchFrom(plan).whereIsInstanceOfAny(SKIP_PLAN_NODES).matches()) {
+            return plan;
+        }
+
+        try {
+            return determineHashPartitionCount(plan, session, types, tableStatsProvider)
+                    .map(hashPartitionCount -> rewriteWith(new Rewriter(hashPartitionCount), plan))
+                    .orElse(plan);
+        }
+        catch (RuntimeException e) {
+            log.warn(e, "Error occurred when determining hash partition count for query %s", session.getQueryId());
+        }
+
+        return plan;
+    }
+
+    private Optional<Integer> determineHashPartitionCount(PlanNode plan, Session session, TypeProvider types, TableStatsProvider tableStatsProvider)
+    {
+        int maxHashPartitionCount = getMaxHashPartitionCount(session);
+        long minInputRowsPerTask = getMinInputRowsPerTask(session);
+
+        if (minInputRowsPerTask == 0) {
+            return Optional.empty();
+        }
+
+        // Skip for expanding plan nodes like CROSS JOIN or UNNEST which can substantially increase the amount of data.
+        if (isExpandingPlanNodesPresent(plan)) {
+            return Optional.empty();
+        }
+
+        CachingStatsProvider statsProvider = new CachingStatsProvider(statsCalculator, session, types, tableStatsProvider);
+
+        double sourceTablesRowCount = getSourcePlanNodesRowCount(plan, statsProvider);
+        double joinNodesMaxRowCount = getJoinPlanNodesMaxRowCount(plan, statsProvider);
```

Comment:
`UnionNode` or `ExchangeNode` with multiple **different** sources can also increase row count

Reply:
We are handling this since in `getSourcePlanNodesRowCount` we are doing `sum` operation